src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/General/Helpers.cs

```diff
@@ -151,22 +151,26 @@ public static BinderBundle ToBinderBundle(this Binder binder, BindingFlags invok
             Justification = "Array.CreateInstance is only used with reference types here and is therefore safe.")]
         public static object[] InstantiateAsArray(this IEnumerable<CustomAttributeData> cads, Type actualElementType)
         {
-            LowLevelList<object> attributes = new LowLevelList<object>();
+            ArrayBuilder<object> attributes = default;
             foreach (CustomAttributeData cad in cads)
             {
                 object instantiatedAttribute = cad.Instantiate();
                 attributes.Add(instantiatedAttribute);
             }
 
-            // This is here for desktop compatibility. ICustomAttribute.GetCustomAttributes() normally returns an array of the
-            // exact attribute type requested except in two cases: when the passed in type is an open type and when
-            // it is a value type. In these two cases, it returns an array of type Object[].
-            bool useObjectArray = actualElementType.ContainsGenericParameters || actualElementType.IsValueType;
-            int count = attributes.Count;
-            object[] result = useObjectArray ? new object[count] : (object[])Array.CreateInstance(actualElementType, count);
-
-            attributes.CopyTo(result, 0);
-            return result;
+            if (actualElementType.ContainsGenericParameters || actualElementType.IsValueType)
+            {
+                // This is here for desktop compatibility. ICustomAttribute.GetCustomAttributes() normally returns an array of the
+                // exact attribute type requested except in two cases: when the passed in type is an open type and when
+                // it is a value type. In these two cases, it returns an array of type Object[].
+                return attributes.ToArray();
+            }
+            else
+            {
+                object[] result = (object[])Array.CreateInstance(actualElementType, attributes.Count);
+                attributes.AsSpan(0).CopyTo(result);
```

Comment:
`CopyTo(Span)` can't be used when destination is covariant. Unlike `CopyTo(Array)` which does type test on each access, span does type test on construction and rejects for covariant case.